### PR TITLE
DAO created for specific entities.

### PR DESCRIPTION
**Pull request by alexgphan**
Currently I've tested pulling information and letting it create instances of the classes from that and it works fine but I have yet to try saving the info from main. I was able to practice saving information from another testfile but not with the current completed ones.

### CODE REVIEW

File: src/app/Main.java [hunk 1]
```diff
@@ -61,21 +67,39 @@ public static void main(String[] args) throws IOException {
...
-        InMemoryEventDataAccessObject eventDAO = dataAccess.getEventDAO();
 
-        // Instantiate EnrollView
-        instantiateEnrollUseCase(employeeDAO, courseDAO, views, viewManagerModel, mySessionsView);
+        // Pull from the employees collection to make in memory employees.
+        FileEmployeeDataAccessObject employeeDataAccessObject;
+        try {
+        employeeDataAccessObject = new FileEmployeeDataAccessObject("./database.csv");
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```
Comment (xubosen) [inline src/app/Main.java:76]: Would it be better if we defined our own exception rather than using a runtime exception?

File: src/app/Main.java [hunk 2]
```diff
@@ -61,21 +67,39 @@ public static void main(String[] args) throws IOException {
...
+        }
+        InMemoryEmployeeDataAccessObject employeeDAO1 = new InMemoryEmployeeDataAccessObject(employeeDataAccessObject.getAccount());
+
+        // Pull from the courses collection to make in memory courses.
+        FileCourseDataAccessObject courseDataAccessObject = new FileCourseDataAccessObject(employeeDAO1);
```
Comment (xubosen) [inline src/app/Main.java:81]: Should not need the "1" after employeeDAO1, since we only have 1.

File: src/app/Main.java [hunk 3]
```diff
@@ -61,21 +67,39 @@ public static void main(String[] args) throws IOException {
...
-        // Load Data (TODO: Replace this with our actual database access once it is implemented.)
-        HardCodedDAO dataAccess = new HardCodedDAO();
-        InMemoryEmployeeDataAccessObject employeeDAO = dataAccess.getEmployeeDAO();
-        InMemoryCourseDataAccessObject courseDAO = dataAccess.getCourseDAO();
-        InMemorySessionDataAccessObject sessionDAO = dataAccess.getSessionDAO();
-        InMemoryEventDataAccessObject eventDAO = dataAccess.getEventDAO();
 
-        // Instantiate EnrollView
-        instantiateEnrollUseCase(employeeDAO, courseDAO, views, viewManagerModel, mySessionsView);
+        // Pull from the employees collection to make in memory employees.
```
Comment (xubosen) [inline src/app/Main.java:71]: Is it possible to split this into a helper method? Our main is getting quite clustered.

File: build.gradle
Comment (xubosen) [inline build.gradle:1]: It is possible to write a documentation for what this file does? I'm a bit confused.

File: src/data_access/FileCourseDataAccessObject.java [hunk 4]
```diff
@@ -1,133 +1,108 @@
...
 
+import com.mongodb.client.*;
 import entity.*;
+import org.bson.Document;
 
 import java.io.*;
-import java.util.HashMap;
-import java.util.LinkedHashMap;
-import java.util.Map;
+import java.util.*;
```
Comment (xubosen) [inline src/data_access/FileCourseDataAccessObject.java:8]: We probably don't need everything in util. Would it be better if we imported just what we used?

File: src/data_access/FileCourseDataAccessObject.java [hunk 5]
```diff
@@ -1,133 +1,108 @@
...
 
+/**
+ * This Class allows us to pull from the courses collection inside our mongodb Database.
+ * It then creates all instances of courses stored inside so it can be saved into memory for the program to retrieve.
+ */
 public class FileCourseDataAccessObject {
-    private final File csvFile;
 
-    private final Map<String, Integer> headers = new LinkedHashMap<>();
+    public static final HashMap<String, Course> courses = new HashMap<>();
```
Comment (xubosen) [inline src/data_access/FileCourseDataAccessObject.java:16]: Should this be private?

File: src/data_access/FileCourseDataAccessObject.java [hunk 6]
```diff
@@ -1,133 +1,108 @@
...
-
-                    courses.put(courseCode, course);
+    /**
+     * Constructs a new file course DAO using information inside the inMemoryEmployeeDataAccessObject and the course
+     * collection on mongodb.
+     * @param inMemoryEmployeeDataAccessObject the DAO that we will pull from to generate the courses with their staff
+     */
+    public FileCourseDataAccessObject(InMemoryEmployeeDataAccessObject inMemoryEmployeeDataAccessObject) {
+
+        String uri = "mongodb+srv://shinyarkeus:KWhU7JJK5BiBcQ0c@projecthr.u9tq0zb.mongodb.net/?retryWrites=true&w=majority";
```
Comment (xubosen) [inline src/data_access/FileCourseDataAccessObject.java:25]: You probably shouldn't include this in the file. Try saving it in another file and then importing it somehow? 

Also, I'm pretty sure this shouldn't be in the repository at all. I think it's better if we each have our private copy of it instead.

File: src/data_access/FileCourseDataAccessObject.java [hunk 7]
```diff
@@ -1,133 +1,108 @@
...
+        try (MongoClient mongoClient = MongoClients.create(uri)) {
+            // database and collection code goes here
+            MongoDatabase db = mongoClient.getDatabase("hrsystem_database");
```
Comment (xubosen) [inline src/data_access/FileCourseDataAccessObject.java:28]: "hrsystem_database" should probably be a constant variable or something you import from a separate file (see comment on saving the uri in another file)

File: src/data_access/FileEmployeeDataAccessObject.java [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+public class FileEmployeeDataAccessObject {
+    private final File databaseFiles;
+    private final Map<String, Integer> headers = new LinkedHashMap<>();
+    public final HashMap<String, Employee> accounts = new HashMap<>();
+
+    public FileEmployeeDataAccessObject(String databasePath) throws IOException {
+        this.databaseFiles = new File(databasePath);
+
+//        databaseFiles = new File(databasePath);
```
Comment (xubosen) [inline src/data_access/FileEmployeeDataAccessObject.java:25]: Is this meant to be commented out? If yes, would it be better if we just deleted it?

File: src/data_access/FileEmployeeDataAccessObject.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+
+
+import java.io.*;
+import java.util.*;
+
+public class FileEmployeeDataAccessObject {
+    private final File databaseFiles;
+    private final Map<String, Integer> headers = new LinkedHashMap<>();
+    public final HashMap<String, Employee> accounts = new HashMap<>();
```
Comment (xubosen) [inline src/data_access/FileEmployeeDataAccessObject.java:20]: Is this supposed to be public?

File: src/data_access/FileEmployeeDataAccessObject.java [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+//
+//            assert header.equals("databaseLink");
+//            String row;
+//            while ((row = reader.readLine()) != null) {
+//                String[] col = row.split(",");
+//                String uri = String.valueOf(col[headers.get("databaseLink")]);
+//            }
+//        }
+
+        String uri = "mongodb+srv://shinyarkeus:KWhU7JJK5BiBcQ0c@projecthr.u9tq0zb.mongodb.net/?retryWrites=true&w=majority";
```
Comment (xubosen) [inline src/data_access/FileEmployeeDataAccessObject.java:39]: Please see previous comment on storing the uri!

File: src/data_access/FileEmployeeDataAccessObject.java [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+        accounts.put(employee.getUID(), employee);
+        this.save();
+    }
+
+    public Employee get(String userID) {
+        return accounts.get(userID);
+    }
+
+    private void save() {
+        String uri = "mongodb+srv://shinyarkeus:KWhU7JJK5BiBcQ0c@projecthr.u9tq0zb.mongodb.net/?retryWrites=true&w=majority";
```
Comment (xubosen) [inline src/data_access/FileEmployeeDataAccessObject.java:73]: Please see previous comment on storing the uri!

File: src/data_access/FileEmployeeDataAccessObject.java [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+        try (MongoClient mongoClient = MongoClients.create(uri)) {
+
+            // database and collection code goes here
+            MongoDatabase db = mongoClient.getDatabase("hrsystem_database");
```
Comment (xubosen) [inline src/data_access/FileEmployeeDataAccessObject.java:77]: Please see previous comment on hard coding things like "hrsystem_database" and "employees"!

File: src/data_access/FileEmployeeDataAccessObject.java [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+    public void save(Employee employee) {
+        accounts.put(employee.getUID(), employee);
+        this.save();
+    }
+
+    public Employee get(String userID) {
+        return accounts.get(userID);
+    }
+
+    private void save() {
```
Comment (xubosen) [inline src/data_access/FileEmployeeDataAccessObject.java:72]: Could this method be split into more helper methods?

File: src/data_access/FileEventDataAccessObject.java [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+        events.put(event.getEventID(), event);
+        this.save();
+    }
+
+    public Event get(String eventID) {
+        return events.get(eventID);
+    }
+
+    private void save() {
+        String uri = "mongodb+srv://shinyarkeus:KWhU7JJK5BiBcQ0c@projecthr.u9tq0zb.mongodb.net/?retryWrites=true&w=majority";
```
Comment (xubosen) [inline src/data_access/FileEventDataAccessObject.java:51]: Please see previous comment on storing the uri!

File: src/data_access/FileEventDataAccessObject.java [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+import java.lang.reflect.Array;
+import java.util.*;
+
+public class FileEventDataAccessObject {
+
+    public static final HashMap<String, Event> events = new HashMap<>();
+    public FileEventDataAccessObject(InMemoryEmployeeDataAccessObject inMemoryEmployeeDataAccessObject,
+                                      InMemoryCourseDataAccessObject inMemoryCourseDataAccessObject) {
+
+        String uri = "mongodb+srv://shinyarkeus:KWhU7JJK5BiBcQ0c@projecthr.u9tq0zb.mongodb.net/?retryWrites=true&w=majority";
```
Comment (xubosen) [inline src/data_access/FileEventDataAccessObject.java:17]: Please see previous comment on storing the uri!

File: src/data_access/FileEventDataAccessObject.java [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+import entity.*;
+import org.bson.Document;
+
+import java.io.*;
+import java.lang.reflect.Array;
+import java.util.*;
+
+public class FileEventDataAccessObject {
+
+    public static final HashMap<String, Event> events = new HashMap<>();
```
Comment (xubosen) [inline src/data_access/FileEventDataAccessObject.java:13]: I don't think this is supposed to be public?

File: src/data_access/FileEventDataAccessObject.java [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+    public void save(Event event) {
+        events.put(event.getEventID(), event);
+        this.save();
+    }
+
+    public Event get(String eventID) {
+        return events.get(eventID);
+    }
+
+    private void save() {
```
Comment (xubosen) [inline src/data_access/FileEventDataAccessObject.java:50]: Could this be split into more helper methods for testing purposes?

File: src/data_access/FileSessionDataAccessObject.java [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+import java.util.Map;
+
+
+// TODO: implement an interface in between so I can follow Clean Architecture
+public class FileSessionDataAccessObject {
+    public static final HashMap<String, ClassSession> sessions = new HashMap<>();
+    public FileSessionDataAccessObject(InMemoryEmployeeDataAccessObject inMemoryEmployeeDataAccessObject,
+                                       InMemoryEventDataAccessObject inMemoryEventDataAccessObject) {
+
+        String uri = "mongodb+srv://shinyarkeus:KWhU7JJK5BiBcQ0c@projecthr.u9tq0zb.mongodb.net/?retryWrites=true&w=majority";
```
Comment (xubosen) [inline src/data_access/FileSessionDataAccessObject.java:21]: Please see previous comment on storing the uri!

File: src/data_access/FileSessionDataAccessObject.java [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+
+// TODO: implement an interface in between so I can follow Clean Architecture
+public class FileSessionDataAccessObject {
+    public static final HashMap<String, ClassSession> sessions = new HashMap<>();
```
Comment (xubosen) [inline src/data_access/FileSessionDataAccessObject.java:17]: Was this meant to be public?

File: src/data_access/FileSessionDataAccessObject.java [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+    public FileSessionDataAccessObject(InMemoryEmployeeDataAccessObject inMemoryEmployeeDataAccessObject,
+                                       InMemoryEventDataAccessObject inMemoryEventDataAccessObject) {
+
+        String uri = "mongodb+srv://shinyarkeus:KWhU7JJK5BiBcQ0c@projecthr.u9tq0zb.mongodb.net/?retryWrites=true&w=majority";
+        // TODO: The uri shouldn't be hardcoded, it should be stored in a file so you can access multiple different databases.
+        // Even in this case where we only have one database it is still recommended.
+        try (MongoClient mongoClient = MongoClients.create(uri)) {
+            // database and collection code goes here
+            MongoDatabase db = mongoClient.getDatabase("hrsystem_database");
```
Comment (xubosen) [inline src/data_access/FileSessionDataAccessObject.java:26]: Please see previous comment on hardcoding "hrsystem_database" and "sessions"!

File: src/data_access/FileSessionDataAccessObject.java [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+        sessions.put(session.getSessionID(), session);
+        this.save();
+    }
+
+    public ClassSession get(String sessionID) {
+        return sessions.get(sessionID);
+    }
+
+    private void save() {
+        String uri = "mongodb+srv://shinyarkeus:KWhU7JJK5BiBcQ0c@projecthr.u9tq0zb.mongodb.net/?retryWrites=true&w=majority";
```
Comment (xubosen) [inline src/data_access/FileSessionDataAccessObject.java:56]: Please see previous comment on storing the uri!

File: src/data_access/FileSessionDataAccessObject.java [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+        try (MongoClient mongoClient = MongoClients.create(uri)) {
+
+            // database and collection code goes here
+            MongoDatabase db = mongoClient.getDatabase("hrsystem_database");
```
Comment (xubosen) [inline src/data_access/FileSessionDataAccessObject.java:60]: Please see previous comment on hardcoding "hrsystem_database" and "sessions"!

File: src/app/Main.java (hunk 2 above)
Comment (alexgphan) [reply src/app/Main.java:81]: Will change this now, was like that because of HardCodedDAO.

File: src/data_access/JsonToCourse.java [hunk 23]
```diff
@@ -0,0 +1,99 @@
...
+
+            // TODO: Consider variable changes.
+
+            return course;
+
+            // Could potentially use course factory if we can recreate the arraylist into Hashmap but this is probably
+            // better because it will have each instance of employee have their corresponding Courses listed underneath.
+        }
+
+    //    public static void main(String[] args) {
```
Comment (xubosen) [inline src/data_access/JsonToCourse.java:63]: Testing code is best deleted when merging.

File: src/data_access/JsonToCourse.java [hunk 24]
```diff
@@ -0,0 +1,99 @@
...
+        }
+
+        /**
+         *
+         * @return
+         */
+        public Course convert() {
+            JSONObject obj = new JSONObject(jsonString);
+
+            String courseName = obj.getString("course_name");
```
Comment (xubosen) [inline src/data_access/JsonToCourse.java:30]: When we rely on how the json file is formatted, we should probably use constants or import them from another file.

File: src/app/Main.java (hunk 3 above)
Comment (alexgphan) [reply src/app/Main.java:71]: Yes, I can split it up into one big helper probably called DAOCreator and then mini helpers inside to create each one.

File: src/data_access/JsonToEmployee.java [hunk 25]
```diff
@@ -0,0 +1,87 @@
...
+
+    public JsonToEmployee(String jsonString) {
+        this.jsonString = jsonString;
+    }
+
+    /**
+     * @return
+     */
+    public Employee convert() {
+//        Employee employee;
```
Comment (xubosen) [inline src/data_access/JsonToEmployee.java:23]: Was this meant to be commented our? If so, would it be better to just delete it?

File: src/data_access/JsonToEmployee.java [hunk 26]
```diff
@@ -0,0 +1,87 @@
...
+
+        JSONObject obj = new JSONObject(jsonString);
+        // "{\"_id\": {\"$oid\": \"6557dd0d9a2a09870f0db0e8\"}, " +
+        //                "\"userID\": \"phanale4\", \"password\": \"123\", \"name\": \"Alexander Phan\", " +
+        //                "\"email\": \"alexanderphan@mail.utoronto.ca\", " +
+        //                "\"courses\": [{\"CSC207\":csc207,\"CSC236\":csc236}], \"sessions\": []}" +
+        //                "\"role\": \"class entity TeachingAssistant\""
+        // test code
+
+        String userID = obj.getString("userID");
```
Comment (xubosen) [inline src/data_access/JsonToEmployee.java:36]: Please see comment on relying on how the json is formatted!

File: src/data_access/JsonToEmployee.java [hunk 27]
```diff
@@ -0,0 +1,87 @@
...
+            TeachingAssistant teachingAssistant = new TeachingAssistant(userID, password, name, email);
+            return teachingAssistant;
+        } else {
+            Instructor instructor = new Instructor(userID, password, name, email);
+            return instructor;
+        }
+
+    }
+}
+//    public static void main(String[] args) {
```
Comment (xubosen) [inline src/data_access/JsonToEmployee.java:52]: Code for testing purposes is best deleted or turned into unit tests.

File: src/data_access/JsonToEvent.java [hunk 28]
```diff
@@ -0,0 +1,91 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+
+
+public class JsonToEvent {
+    private InMemoryEmployeeDataAccessObject inMemoryEmployeeDataAccessObject;
+    private InMemoryCourseDataAccessObject inMemoryCourseDataAccessObject;
+    private String jsonString;
+
+//    public static void main(String[] args) {
```
Comment (xubosen) [inline src/data_access/JsonToEvent.java:18]: Code for testing purposes is best deleted or turned into unit tests.

File: src/data_access/JsonToEvent.java [hunk 29]
```diff
@@ -0,0 +1,91 @@
...
+    public Event convert() {
+        JSONObject obj = new JSONObject(jsonString);
+        // "{\"_id\": {\"$oid\": \"6557dd0d9a2a09870f0db0e8\"}, " +
+        //                "\"userID\": \"phanale4\", \"password\": \"123\", \"name\": \"Alexander Phan\", " +
+        //                "\"email\": \"alexanderphan@mail.utoronto.ca\", " +
+        //                "\"courses\": [{\"CSC207\":csc207,\"CSC236\":csc236}], \"sessions\": []}" +
+        //                "\"role\": \"class entity TeachingAssistant\""
+        // test code
+
+        String name = obj.getString("name");
```
Comment (xubosen) [inline src/data_access/JsonToEvent.java:59]: Please see comment on relying on how the json file is formatted!

File: src/data_access/JsonToSession.java [hunk 30]
```diff
@@ -0,0 +1,91 @@
...
+     *
+     * @return
+     */
+    public ClassSession convert() {
+        // TODO: Convert all of the info to the session collection info.
+        JSONObject obj = new JSONObject(jsonString);
+
+        // The plan is to store the name description and the date time span into a arrayList
+
+        String sessionName = obj.getString("session_name");
```
Comment (xubosen) [inline src/data_access/JsonToSession.java:34]: Please see comment on relying on how the json file is formatted!

File: src/data_access/JsonToEvent.java [hunk 31]
```diff
@@ -0,0 +1,91 @@
...
+        this.jsonString = jsonString;
+        this.inMemoryEmployeeDataAccessObject = inMemoryEmployeeDataAccessObject;
+        this.inMemoryCourseDataAccessObject = inMemoryCourseDataAccessObject;
+
+    }
+
+    /**
+     * @return
+     */
+    public Event convert() {
```
Comment (xubosen) [inline src/data_access/JsonToEvent.java:50]: Could this be split into more helper methods?

File: src/data_access/JsonToSession.java [hunk 32]
```diff
@@ -0,0 +1,91 @@
...
+        this.jsonString = jsonStringSession;
+        this.inMemoryEmployeeDataAccessObject = inMemoryEmployeeDataAccessObject;
+        this.inMemoryEventDataAccessObject = inMemoryEventDataAccessObject;
+    }
+
+    /**
+     *
+     * @return
+     */
+    public ClassSession convert() {
```
Comment (xubosen) [inline src/data_access/JsonToSession.java:28]: This method is best split into several helper methods.

File: src/entity/Employee.java
Comment (xubosen) [inline src/entity/Employee.java:1]: Is this meant to be commited?

File: src/data_access/FileCourseDataAccessObject.java (hunk 4 above)
Comment (alexgphan) [reply src/data_access/FileCourseDataAccessObject.java:8]: I should because it would more clear what we are using from the library.

**Review by xubosen — CHANGES_REQUESTED** [review]
Please see more details for in-line comments! 

Also, since these are rather complicated classes, there should probably be unit tests testing most if not all of the helper methods of each of these classes, just to make sure they work.

File: src/entity/Employee.java
Comment (alexgphan) [reply src/entity/Employee.java:1]: No, it was just a really outdated todo.

**Review by xubosen — COMMENTED** [review]
Another note is that currently this Data Access implementation doesn't abide by clean architecture. 

I think a way to edit this is to split the current "DataAccessObject" classes into a "data access" version that just retrieves the JSON objects from mongoDB and an "interactor" version that takes those strings and create the actual entities.

(You could even turn the JSON objects into an intermediate data package if you like, similar to how we use input data and output data for use cases)

**Review by xubosen — APPROVED** [review]
Clean architecture problem still exists, but most previous issues are fixed. Testing for data access is complete. Due to time constraints, merging is done although content is yet imperfect. More fixes to be done.

**Pull request merged**